src/eradiate/scenes/surface/_dem.py

```diff
@@ -0,0 +1,250 @@
+
+        return face_indices
+
+    @classmethod
+    def from_dataarray(
+        cls, data: xr.DataArray, bsdf: BSDF, planet_radius=6371 * ureg.km
+    ):
+        """
+        This classmethod enables the instantiation of a DEM from a dataarray holding elevation data.
+        The dataarray must use latitude and longitude as its dimensional coordinates.
+
+        Parameters
+        ----------
+        data : xarray.dataarray
+            Dataarray holding the elevation data. Dimensional coordinates must be latitude `lat`
+            and longitude `lon`. Data will be interpreted to be given in kernel units.
+
+        bsdf : .BSDF
+            BSDF to be attached to the mesh shape.
+
+        planet_radius : quantity
+            Planet radius used to convert latitude and longitude into distance units.
+            Defaults to earth's radius.
+
+        Returns
+        -------
+        :class:`.DEMSurface`
+            Created :class:`DEMSurface`.
+        """
+        radius = planet_radius.m_as(ucc.get("length"))
+
+        len_lat = len(data.lat)  # x
+        len_lon = len(data.lon)  # y
+
+        # I convert the degrees to local distances based on the planet radius
+        # This will be handled differently when I include the AtmosphereGeometry.
+        mean_lon = data.lon.mean()
+        lon_range = data.lon.max() - data.lon.min()
+        lat_range = data.lat.max() - data.lat.min()
+        lon_length = np.deg2rad(lon_range) * radius
+        lat_length = np.deg2rad(lat_range) * radius * np.cos(np.deg2rad(mean_lon))
+
+        lat_range = list(np.linspace(-lat_length / 2.0, lat_length / 2.0, len_lat))
+        # Duplicate the first and last entry for the extra vertices, which form the
+        # vertical walls
+        lat_range.insert(0, lat_range[0])
+        lat_range.append(lat_range[-1])
+
+        lon_range = list(np.linspace(-lon_length / 2.0, lon_length / 2.0, len_lon))
+
+        vertex_positions = []
+        values = np.array(data.data).transpose()[:, ::-1]
+        for i, x in enumerate(lat_range):
+            if i in [0, len_lat + 1]:
+                # For the rows of vertices which are only the lower points of the wall
+                # set the z coordinate to -1
+                z_list = np.full(len_lon + 2, -1.0)
+            else:
+                z_list = values[i - 1]
+                # Add z=-1 on both ends for the points that make up the vertical wall
+                z_list = np.insert(z_list, 0, -1.0)
+                z_list = np.append(z_list, -1.0)
+            for j, z in enumerate(z_list):
+                # Duplicate the first and last entry for the extra vertices, which form the
+                # vertical walls
+                if j == 0:
+                    y = lon_range[0]
+                elif j == len_lon + 1:
+                    y = lon_range[-1]
+                else:
+                    y = lon_range[j - 1]
+                vertex_positions.append([x, y, z])
```

Comment:
As in the case of the `from_analytical()` constructor, allocating memory once and for all is easy and cleaner IMO: you know the size of your container in advance, so using structures with dynamic size is not really appropriate. In addition, this is better in the prospect of replacing the loops with vectorised operations for optimal performance.